Impl signing for minting/burning in new tx flow

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->


- [x] cleaning old minting/burning enpoint
- [x] add postPolicyKey endpoint used when the already existent wallet does not have policy xpub
- [x] add signing
- [x] Adjust fee calculation
- [x] show signing in integration testing
- [x] add optional param to POST policy key
- [x] unify construct/decode mint/burn responses 

### Comments
Minting/burning working without timelocks at this moment - needs validation integral fixing to enable this - coming in the next PR
<!-- Additional comments, links, or screenshots to attach, if any. -->
I want to have construct/decode responses:
```
{
  "tokens": [
    {
      "policy_id": "0fc891fb7368d3b7c7b88815c203fda0d6862b0f1d797222672e91fe",
      "policy_script": "policy_vkh1ecqfpeunygcgl4tqqycn86x6mhna99vsrm5pk22vq8dvxfq3htn", 
      "assets" : [
           { "asset_name": "1111",
             "quantity": 1,
             "fingerprint": "asset17jwl24z89xcgljv57fe4de558zvwl6k9h0hf22" 
           }, 
           { "asset_name": "2222",
             "quantity": 10,
             "fingerprint": "asset17jwl24z89xcgljv57fe4de558zvwl66666666" 
           }, 
      ]
    }
  ],
  "wallet_policy_key_index": "0H",
  "wallet_policy_key_hash": "policy_vkh1ecqfpeunygcgl4tqqycn86x6mhna99vsrm5pk22vq8dvxfq3htn"
}
```
### Issue Number
adp-955
adp-1542


<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
clean not needed bits - only new tx flow support minting/burning

paweljakubas committed:
add ApiPostPolicyKeyData

paweljakubas committed:
ApiPostPolicyKeyData - part 2

paweljakubas committed:
add writePolicyKey in Cardano.Wallet

paweljakubas committed:
conclude postPolicyKey endpoint

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -6096,6 +6045,24 @@ paths:
   /wallets/{walletId}/policy-key:
+    post:
+      operationId: postPolicyKey
+      tags: ["Transactions"]
```
I think it should go to "Keys"

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -6096,6 +6045,24 @@ paths:
+    post:
+      operationId: postPolicyKey
+      tags: ["Transactions"]
+      summary: Create
```
```suggestion
      summary: Create Policy Key
```

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -6096,6 +6045,24 @@ paths:
+        <p align="right">status: <strong>stable</strong></p>
+
+        Create policy key for the wallet. Used for public account key derived
+        wallets or already instantiated mnemonic wallets to set policy public key
```
```suggestion
        Create policy key for the wallet. 
        
        In order to be able to mint/burn assets with `POST Construct` endpoint there needs to be 
        a policy key set for the wallet. Invoking this endpoint would be required for all wallets instantiated
        before introducing mint/burn feature prior to making a mint/burn transaction from them.

```

piotr-iohk replied:
I guess that this endpoint now invalidates https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getPolicyKey... I mean there's basically no need for it, right?

Also `postPolicyKey` returns actually a hash (`policy_vkh`), perhaps we should add a query parameter hash=true here and return `policy_vk` or `policy_vkh` based on it?

paweljakubas replied:
now, for the consistency I added optional parameter like in getPolicyKey.
I would advocate for maintaining `getPolicyKey` in line with `postPolicyKey`. Could be helpful for consumers of the API, also POST writes to DB, and GET is reading

piotr-iohk replied:
Ok, sounds good to leave `getPolicyKey`, but perhaps for `postPolicyKey` the `POST` should be actually `PUT`... it doesn't create new one every time, but sets one if doesn't exists (or overwrites the old one, but it is always the same if I'm not mistaken)

paweljakubas committed:
add signing for minting/burning

paweljakubas committed:
extend integration test - part 1

paweljakubas committed:
fix fee calculation

paweljakubas committed:
reorganize integration tests for better reuse

paweljakubas committed:
prove that minted assets are in wallet finally

paweljakubas committed:
prove burning in integration testing

paweljakubas committed:
postPolicyPublicKey with optional param plus cleaning

paweljakubas committed:
unify mint/burn response for construct/decode - swagger

paweljakubas committed:
finish refactoring construct/decode

jonathanknowles committed:
Wrap long lines according to team coding standard.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles committed:
Use just a single blank line between definitions.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

jonathanknowles committed:
Use 4-space indents in Haskell source files.

See `.editorconfig`.

sevanspowell reviewed (commented):
Thanks for this work @paweljakubas. This kind of work which involves making changes from the lowest level to the highest is difficult.

I've read through everything and it seems sensible. However, just like the last PR I reviewed it is hard to give proper review. 

For me, being able to read and understand the tests is more important than the code and it's performance characteristics. In my review I can't assert that:
  - All edge cases are being tested.
  - Tests do indeed test what they're trying to test.

I have two requests:
1. Is it possible to structure the code so that it can be property tested? I'd be really interested in property tests that cover the edge cases of common attack vectors like burning tokens you don't have, minting and burning at the same time, etc.
    - Structuring the code such that it can be property tested may make it easier to review too.
2. Could we raise the level of abstraction in the integration tests? It's becoming increasingly difficult to untangle the core of the integration tests from the code (i.e. "with this input, I expect this output").
    - Even just some high-level functions (no need to make a new DSL) could help here.

I understand you're just working with what you've got, and these aren't issues you've introduced, but I can't give meaningful review without a clearer relationship between the code and the proof that code works.

jonathanknowles commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -3205,66 +3215,19 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
-            ]
+    it "TRANS_NEW_CREATE_10e - Minting assets with timelocks" $
+        \ctx -> runResourceT $ do
+        liftIO $ pendingWith "Should work when interval validity is addressed"
```
@paweljakubas 

Is it possible to add a "TODO" comment here, to link to the issue in JIRA where the internal validity issue is described?

paweljakubas replied:
yes, we have validity interval ticket for that. It is the subject of third step to conclude minting/burning thorough support. I should start doing this next week

jonathanknowles replied:
Okay, I've added a TODO that links to https://input-output.atlassian.net/browse/ADP-1193 at the top of this test!

Fixed in: 230d8c1b56538b6063fd45d7c4545995a317bbea.

jonathanknowles committed:
Add type signatures to helper functions in `TransactionsNew`.

jonathanknowles committed:
Rename `mintAssetCheck` to `mintAssetsCheck`.

This is for consistency with its sister function `burnAssetsCheck`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2474,19 +2446,49 @@ decodeTransaction ctx (ApiT wid) (ApiSerialisedTransaction (ApiT sealed)) = do
+            []
+        else
+           map fromIdScriptAssets $
+           toIdScriptAssets scriptMap tokenMap
```
Just curious if this would also work:

```suggestion
        map fromIdScriptAssets $
        toIdScriptAssets scriptMap tokenMap
```

jonathanknowles replied:
Fixed in 1c4da7cd740b501655847f312ad170f91d345915.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2474,19 +2446,49 @@ decodeTransaction ctx (ApiT wid) (ApiSerialisedTransaction (ApiT sealed)) = do
+            uncurry ApiPolicyKey . computeKeyPayload (Just True) <$>
+            includePolicyKeyInfo tokenWithScripts xpubM
+        , walletPolicyKeyIndex =
+            (const policyIx) <$> includePolicyKeyInfo tokenWithScripts xpubM
```
```suggestion
            policyIx <$ includePolicyKeyInfo tokenWithScripts xpubM
```

jonathanknowles replied:
Fixed in 8c03c468cbf56a891d7b5c229ba9e468dcfc9260.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3362,15 +3364,22 @@ getPolicyKey ctx (ApiT wid) hashed = do
+postPolicyKey ctx (ApiT wid) hashed apiPassphrase =
+    withWorkerCtx @_ @s @ShelleyKey ctx wid liftE liftE $ \wrk -> do
+        k <- liftHandler $ W.writePolicyPublicKey @_ @s @n wrk wid pwd
+        pure $ uncurry ApiPolicyKey (computeKeyPayload hashed (getRawKey k))
```
You've already got one `$`, so perhaps go all the way? 😄 

```suggestion
        pure $ uncurry ApiPolicyKey $ computeKeyPayload hashed $ getRawKey k
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4384,13 +4389,18 @@ instance IsServerError ErrReadPolicyPublicKey where
-                , "transaction. Please restore the wallet from a mnemonic"
-                , "phrase instead of an account public key."
+                , "transaction. Please invoke POST call on endpoint"
+                , "/wallets/{walletId}/policy-key to set it."
```
```suggestion
                , "transaction. Please POST to endpoint"
                , "/wallets/{walletId}/policy-key to set it."
```

jonathanknowles replied:
Fixed in 2a92320c1a0506000d22213a01d2ce8b71387cea.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1223,17 +1219,32 @@ data ApiAnyCertificate n =
 
-data ApiPolicyScript = ApiPolicyScript
+newtype ApiPostPolicyKeyData = ApiPostPolicyKeyData
+    { passphrase :: (ApiT (Passphrase "raw"))
```
```suggestion
    { passphrase :: ApiT (Passphrase "raw")
```

jonathanknowles replied:
Fixed in 7076da5d9e70a8337a0ad7c822c2e49866714431.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1245,8 +1256,8 @@ data ApiDecodedTransaction (n :: NetworkDiscriminant) = ApiDecodedTransaction
-    , assetsMinted :: !ApiAssetMintBurn
-    , assetsBurned :: !ApiAssetMintBurn
+    , mint :: !ApiAssetMintBurn
+    , burn :: !ApiAssetMintBurn
```
I like this: it's much shorter, but still clear! 👍🏻

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1955,6 +1946,9 @@ instance Arbitrary (Quantity "percent" Double) where
     arbitrary = Quantity <$> choose (0,100)
 
+instance Arbitrary ApiPostPolicyKeyData where
+    arbitrary = ApiPostPolicyKeyData <$> arbitrary
```
You can save a bit of boilerplate here:
```suggestion
deriving instance Arbitrary ApiPostPolicyKeyData
```

jonathanknowles replied:
Fixed in ba75be5e521293bb6adeb392c92fb52f0c625a52.

jonathanknowles committed:
Rename `ApiTokenAsset` to `ApiTokenAmountFingerprint`.

After review discussion.

jonathanknowles committed:
Simplify applicative expressions.

(Remove redundant `<$> pure` combinations.)

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1223,17 +1219,32 @@ data ApiAnyCertificate n =
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiTokenAsset = ApiTokenAsset
```
Perhaps call this `ApiTokenAmountFingerprint`?

I feel the juxtaposition of `Token` and `Asset` might be confusing here.

(And it there's the fact that this also includes an amount and a fingerprint.)

jonathanknowles replied:
Fixed in 6951314468399a163fa58739c5b036bf1cbcb3b4.

jonathanknowles committed:
Use `GeneralizedNewtypeDeriving` to eliminate boilerplate.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -463,6 +472,16 @@ signTransaction networkId resolveRewardAcct resolveAddress resolveInput (body, w
+        mintBurnScripts =
+            let (_, toMint, toBurn, _) = fromCardanoTx $
+                    Cardano.makeSignedTransaction wits body
+            in L.nub $ getScripts toMint <> getScripts toBurn
```
Question: why is it important to use `L.nub` here?

paweljakubas replied:
because, we can have scripts having the same policyXPub repeated and it is enough to deliver ONE wits for that

paweljakubas replied:
example:
```
any [ all [cosigner#0, after slot1], all [cosigner#0, before slot2]
```
One wits is needed here. the one produced with signing key for cosigner#0. See it?

jonathanknowles replied:
Okay, I've added a comment to the code:

90499017b73e990491cd7dbf0ffc735ee828955c

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -477,6 +496,10 @@ signTransaction networkId resolveRewardAcct resolveAddress resolveInput (body, w
 
+    mkPolicyWitness :: KeyHash -> Maybe (Cardano.KeyWitness era)
+    mkPolicyWitness a = do
+        mkShelleyWitness body <$> resolvePolicyKey a
```
```suggestion
    mkPolicyWitness a =
        mkShelleyWitness body <$> resolvePolicyKey a
```

jonathanknowles replied:
Fixed in 7076da5d9e70a8337a0ad7c822c2e49866714431.

jonathanknowles committed:
Rename `numberOfAssets` to `tokenAmount`.

This is consistent with the `ApiTokenAmountFingerprint` name.

jonathanknowles reviewed (commented):
Hi @paweljakubas 

I've left some comments and questions!

jonathanknowles committed:
Revise error message for `MissingPolicyPublicKey`.

jonathanknowles committed:
Add note regarding the use of `L.nub`.

See review conversation here:

https://github.com/input-output-hk/cardano-wallet/pull/3199#discussion_r839398755

jonathanknowles committed:
Add TODO that links to ADP-1193 in `TRANS_NEW_CREATE_10e`.

jonathanknowles committed:
Replace `const x <$>` with `x <$` in `Api.Server`.

jonathanknowles committed:
Remove redundant parentheses and `do` usages.

jonathanknowles committed:
Remove redundant check for empty `TokenMap` in `toApiTokens`.

jonathanknowles reviewed (approved):
Hi Pawel

Many thanks for creating this PR!

I think it's going in the right direction. However, I do share the concerns that @sevanspowell  expressed above, namely that:

> Is it possible to structure the code so that it can be property tested? I'd be really interested in property tests that cover the edge cases of common attack vectors like burning tokens you don't have, minting and burning at the same time, etc.

I also feel we should invest more time in writing some focussed property tests, in addition to the integration tests you've already written.

I feel the main advantage would be that we can generate hundreds of random examples, rather than the very small number we're able to test in our integration test framework. We'd also gain a shorter editing-testing iteration cycle, which means faster feedback when things break. Lastly, it might help us to break code up into smaller, clearer chunks, which would answer Sam's point about making the code easier to review.

At the end of the day, as you pointed out, we do still need the integration tests to prove that the system works from end-to-end, but I feel this should be more of a "sanity check" rather than the main thrust of our test effort.

Perhaps we can add more property testing in the next PR? In particular, I'm curious to verify what happens in the following edge cases, in addition to what Sam already mentioned:

- minting and burning in the same transaction (disjoint sets of assets)
- minting and burning in the same transaction (intersecting sets of assets: this might provoke some failures)
- attempting to burn assets that are not available
- attempting to mint or burn large numbers of assets (so that the map is very large: I wonder whether we'll hit upon some kind of limit).

I suspect there are also lots of other things we can test.

Anyway, I'm happy to approve, but I do think it would also be good to also get a review from someone who is more knowledgeable about what kind of design we really want for the new transaction API. (I can only review this part very superficially!) Perhaps @rvl has an opinion here.

Thanks again!

Jonathan
